store/src/main/java/com/book/store/service/impl/BookServiceImpl.java

```diff
@@ -66,5 +69,47 @@ public void deleteBook(Long id) {
             throw new EntityNotFoundException("Book with id " + id + " not found.");
         }
     }
-}
 
+    @Override
+    public List<BookDto> searchBooks(
+            String title,
+            String author,
+            BigDecimal price,
+            String description
+    ) {
+        return bookRepository
+                .findAll((root, query, criteriaBuilder) -> {
+                    List<Predicate> predicates = new ArrayList<>();
+
+                    if (title != null && !title.isEmpty()) {
+                        predicates.add(criteriaBuilder
+                                .like(criteriaBuilder
+                                        .lower(root.get("title")), "%"
+                                        + title.toLowerCase() + "%"));
+                    }
+
+                    if (author != null && !author.isEmpty()) {
+                        predicates.add(criteriaBuilder
+                                .like(criteriaBuilder
+                                        .lower(root.get("author")), "%"
+                                        + author.toLowerCase() + "%"));
+                    }
+
+                    if (price != null) {
+                        predicates.add(criteriaBuilder
+                                .equal(root.get("price"), price));
+                    }
+
+                    if (description != null && !description.isEmpty()) {
+                        predicates.add(criteriaBuilder
+                                .like(criteriaBuilder
+                                        .lower(root.get("description")), "%"
+                                        + description.toLowerCase() + "%"));
+                    }
+
+                    return criteriaBuilder.and(predicates.toArray(new Predicate[0]));
+                }).stream()
+                .map(bookMapper::toBookDto)
+                .toList();
```

Comment:
ugly enormous lambda expression.

Extract it at separate class, always remember about SRP.